packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,261 @@
+import { NodePHP } from '@php-wasm/node';
+
+let childCount = 0;
+
+/**
+ * PRIVATE
+ * Tracks stats of instances in a pool.
+ */
+class PoolInfo {
+	id = childCount++; // Unique ID for debugging purposes.
+	requests = 0; // Total requests processed.
+	started = Date.now(); // Time spawned.
+	active = false; // Whether instance is considered active.
+}
+
+/**
+ * PRIVATE
+ * Spawns new instances if the pool is not full.
+ * Returns a list of new instances.
+ */
+const spawn = (pool) => {
+	const newInstances = new Set();
+
+	while (pool.maxJobs > 0 && pool.instanceInfo.size < pool.maxJobs) {
+		const info = new PoolInfo();
+		const instance = pool.spawner();
+		pool.instanceInfo.set(instance, info);
+		info.active = true;
+		newInstances.add(instance);
+	}
+
+	return newInstances;
+};
+
+/**
+ * PRIVATE
+ * Reaps children if they've passed the maxRequest count.
+ */
+const reap = (pool) => {
+	for (const [instance, info] of pool.instanceInfo) {
+		if (pool.maxRequests > 0 && info.requests >= pool.maxRequests) {
+			info.active = false;
+			pool.instanceInfo.delete(instance);
+			instance.then((unwrapped) => unwrapped.exit()).catch((error) => {});
+			continue;
+		}
+	}
+};
+
+/**
+ * PRIVATE
+ * Handle fatal errors gracefully.
+ */
+const fatal = (pool, instance, error) => {
+	console.error(error);
+
+	if (pool.instanceInfo.has(instance)) {
+		const info = pool.instanceInfo.get(instance);
+		info.active = false;
+		pool.instanceInfo.delete(instance);
+	}
+};
+
+/**
+ * PRIVATE
+ * Find the next available idle instance.
+ */
+const getIdleInstance = (pool) => {
+	const sorted = [...pool.instanceInfo].sort(
+		(a, b) => a[1].requests - b[1].requests
+	);
+
+	for (const [instance, info] of sorted) {
+		if (pool.running.has(instance)) {
+			continue;
+		}
+
+		if (!info.active) {
+			continue;
+		}
+		return instance;
+	}
+
+	return false;
+};
+
+/**
+ * Maintains and refreshes a list of php instances
+ * such that each one will only be fed X number of requests
+ * before being discarded and replaced.
```

Comment:
https://www.php.net/manual/en/install.fpm.configuration.php#pm.max-requests

This pathological behavior does exist in native PHP, but won't cause problems nearly as quickly since its not restricted to a single gigabyte of memory. 

Emscripten definitely introduces unique challenges. Since we're dealing with a linear, "physical" memory array, we're prone to things like memory fragmentation. In that situation, we could have the entire gigabyte empty except for a few sparse allocations. If no **_contiguous_** region of memory exists for the length requested, memory allocations will fail. This tends to happen when a new request attempts to initialize a heap structure but cannot find a contiguous  2mb chunk of memory.

We can go as far as debugging PHP itself, and contributing the fix upstream. But even in this case we cannot guarantee that a third party extension will not introduce a leak sometime in the future.

Therefore, we should have a solution robust to memory leaks that come from upstream code. I think that following the native strategy is the best way.

Reply:
Oh of course, third party extensions! Sean, that's such a good answer, thank you. I'd love to see your GitHub comment as a part of this PR in an actual comment block – it clearly explains the rationale.

> We can go as far as debugging PHP itself, and contributing the fix upstream. But even in this case we cannot guarantee that a third party extension will not introduce a leak sometime in the future.

Agreed, let's just stick with discarding PHP instances here.

Reply:
@adamziel  https://github.com/WordPress/playground-tools/pull/110/files#diff-95251f05603f485be0377cd57ce90fe740821ca840c6053a115cb48c8725341bR100

Reply:
https://github.com/WordPress/playground-tools/blob/f352bb84c1c67a45bbb27434c19b463f7a33b908/packages/wp-now/src/pool.ts#L110-L131